src/hidden.rs

```diff
@@ -20,74 +20,248 @@
-pub struct Hidden<T> {
-    inner: T,
+pub struct Hidden<T>
+where
+    T: Zeroize,
+{
+    inner: Box<T>,
 }
 
-impl<T> Hidden<T> {
-    /// Hides a value.
+impl<T> Hidden<T>
+where
+    T: Zeroize,
+{
+    /// Create new hidden data from the underlying type
     pub fn hide(inner: T) -> Self {
-        Self { inner }
-    }
-
-    /// Returns ownership of the inner value discarding the wrapper.
-    pub fn into_inner(self) -> T {
-        self.inner
+        Self {
+            inner: Box::new(inner),
+        }
     }
 
-    /// Provides access to the inner value (immutable).
+    /// Reveal the hidden data as an immutable reference
     pub fn reveal(&self) -> &T {
         &self.inner
     }
 
-    /// Provides access to the inner value (mutable).
+    /// Reveal the hidden data as a mutable reference
     pub fn reveal_mut(&mut self) -> &mut T {
         &mut self.inner
     }
 }
 
-impl<T> From<T> for Hidden<T> {
-    fn from(inner: T) -> Self {
-        Self::hide(inner)
+/// Only output masked data for debugging, keeping the hidden data hidden
+impl<T> fmt::Debug for Hidden<T>
+where
+    T: Zeroize,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Hidden<{}>", type_name::<T>())
     }
 }
 
-/// Defines a masked value for the type to output as debug information. Concealing the secrets within.
-impl<T> fmt::Debug for Hidden<T> {
+/// Only display masked data, keeping the hidden data hidden
+impl<T> fmt::Display for Hidden<T>
+where
+    T: Zeroize,
+{
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Hidden<{}>", std::any::type_name::<T>())
+        write!(f, "Hidden<{}>", type_name::<T>())
     }
 }
 
-/// Defines a masked value for the type to display. Concealing the secrets within.
-impl<T> fmt::Display for Hidden<T> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Hidden<{}>", std::any::type_name::<T>())
+/// Zeroize the hidden data
+impl<T> Zeroize for Hidden<T>
+where
+    T: Zeroize,
+{
+    fn zeroize(&mut self) {
+        self.inner.zeroize();
+    }
+}
+
+/// Zeroize the hidden data when dropped
+impl<T> Drop for Hidden<T>
+where
+    T: Zeroize,
+{
+    fn drop(&mut self) {
+        self.zeroize();
     }
 }
 
 #[cfg(test)]
-mod test {
+mod tests {
+    use rand::rngs::OsRng;
+    use rand::RngCore;
+
     use super::*;
 
     #[test]
-    fn into_applies_wrapper_deref_removes_it() {
-        let wrapped: Hidden<u8> = Hidden::hide(42);
-        assert_eq!(42, *wrapped.reveal());
+    fn references() {
+        // Check immutable reference
+        assert_eq!(Hidden::hide(0u8).reveal(), &0u8);
+
+        // Check mutable reference
+        let mut hidden = Hidden::hide([0u8; 32]);
+        let mut rng = OsRng;
+        rng.fill_bytes(hidden.reveal_mut());
+    }
+
+    #[test]
+    fn serialize() {
+        let hidden = Hidden::<u8>::hide(1u8);
+        let ser = serde_json::to_string(&hidden).unwrap();
```

Comment:
We should disallow this from a semantic point of view, since this will be a source of accidental leaks